private/buf/bufgen/features.go

```diff
@@ -22,86 +22,189 @@ import (
-				features[feature] = append(features[feature], imageFile.Path())
+				features.flags[feature] = append(features.flags[feature], imageFile.Path())
 			}
 		}
+		// We also collect the range of required editions.
+		if !fileHasEditions(imageFile.FileDescriptorProto()) {
+			continue
+		}
+		edition := imageFile.FileDescriptorProto().GetEdition()
+		features.editions[edition] = append(features.editions[edition], imageFile.Path())
+		if features.minEdition == 0 || edition < features.minEdition {
+			features.minEdition = edition
+		}
+		if edition > features.maxEdition {
+			features.maxEdition = edition
+		}
 	}
 	return features
 }
 
 func checkRequiredFeatures(
 	container app.StderrContainer,
-	required requiredFeatures,
+	required *requiredFeatures,
 	responses []*pluginpb.CodeGeneratorResponse,
 	configs []bufconfig.GeneratePluginConfig,
-) {
+) error {
+	var failedPlugins []string
 	for responseIndex, response := range responses {
 		if response == nil || response.GetError() != "" {
 			// plugin failed, nothing to check
 			continue
 		}
-		failed := requiredFeatures{}
+
+		failed := newRequiredFeatures()
 		var failedFeatures []pluginpb.CodeGeneratorResponse_Feature
+		var failedEditions []descriptorpb.Edition
 		supported := response.GetSupportedFeatures() // bit mask of features the plugin supports
-		for feature, files := range required {
-			featureMask := (uint64)(feature)
+		for feature, files := range required.flags {
+			featureMask := uint64(feature)
 			if supported&featureMask != featureMask {
 				// doh! Supported features don't include this one
-				failed[feature] = files
+				failed.flags[feature] = files
 				failedFeatures = append(failedFeatures, feature)
 			}
 		}
-		if len(failed) > 0 {
-			// TODO: plugin config to turn this into an error
-			_, _ = fmt.Fprintf(container.Stderr(), "Warning: plugin %q does not support required features.\n",
-				configs[responseIndex].Name())
+		if supported&uint64(pluginpb.CodeGeneratorResponse_FEATURE_SUPPORTS_EDITIONS) != 0 && len(required.editions) > 0 {
+			// Plugin supports editions, and files include editions. So make sure
+			// the plugin supports precisely the right editions.
+			requiredEditions := make([]descriptorpb.Edition, 0, len(required.editions))
+			for edition := range required.editions {
+				requiredEditions = append(requiredEditions, edition)
+			}
+			sort.Slice(requiredEditions, func(i, j int) bool {
+				return requiredEditions[i] < requiredEditions[j]
+			})
+			for _, requiredEdition := range requiredEditions {
+				if int32(requiredEdition) < response.GetMinimumEdition() ||
+					int32(requiredEdition) > response.GetMaximumEdition() {
+					failed.editions[requiredEdition] = required.editions[requiredEdition]
+					failedEditions = append(failedEditions, requiredEdition)
+				}
+			}
+		}
+
+		pluginName := configs[responseIndex].Name()
+		if len(failedFeatures) > 0 {
+			_, _ = fmt.Fprintf(
+				container.Stderr(),
```

Comment:
Can you explain why we're printing to stderr here (and everywhere else) instead of returning an error?

Reply:
I called this out in the PR description -- about whether to log details to stderr or just stuff everything into the returned error. Based on this comment, it sounds like you'd prefer to see it all stuffed into the error.

Reply:
Changed